### ISSUE TITLE
Widgets should define somewhere the needed roles to access it

### ISSUE THREAD

**Issue by adam-buckley** [2pisoftware/cmfive#122]
Currently all widgets are showing, the list/viewing of which should be controlled by roles

**Comment by adam-buckley** [comment]
@careck I'm not overly sure how to go about this. Ideally we would define a minimum set of roles that a user needs to even see the widget in a list. 

I.e having a widget that shows 10 most recent unread would require the user to have the inbox_reader role at least

The widgets use an abstract class so we could define in code somewhere what roles it needs to display...
What do you think?

**Comment by adam-buckley** [comment]
@careck bump

**Comment by adam-buckley** [comment]
There's a new function in the ModuleWidget class called getRequiredRoles that the widget ajax action uses to strip out widgets that the user doesn't have the required role for. If empty, the widget will show for everyone.

If we want a widget to be restricted, in the widget class we override the getRequiredRoles function and set the array of required roles:
```php
...
public function getRequriedRoles() {
    return ["comment", "task_user"];
}
```

And the widget will only be available for selection to those who have the specified roles.

**Comment by careck** [comment]
This is not really "the cmfive way" in that it ties the appearance of a widget too specifically to a role. What if a widget should only appear if the user "has been assigned to a business opportunity" or some such business logic?

I suggest to mimic the way access is controlled on objects using a canView() method and inside that method we could then ask for the role, eg.:
`
public function canView(User $user) {
    return $user->hasAnyRole(["comment","task"]);
}
`

**Comment by adam-buckley** [comment]
Fixed in 129031e5c4ebb889cdd70b76b0ef30b7dbf87286
